### DISCUSSION TITLE
Generate method

### DISCUSSION THREAD

**Discussion by BladeMF** [phpactor/phpactor#1255] [General]
Strategy for implementing a "generate method" transform:

## General
* It's a `code_transform.transformer`.
* Provides both diagnostics and a code action

## Diagnostics
1. Find all method calls (__using the MS parser__)
2. Reflect each call  (__supposedly with `Reflector::reflectMethod`__) and determine:
    1) Is the class in the current workspace? (__not sure how__)
    2) Is the method unresolved (__not sure how__)
3. If both yes, add a diagnostic

## Action
1. Call `GenerateMethod::generateMethod`

**Comment by BladeMF** [comment]
I'd appreciate help with the reflecting stuff.

**Reply by dantleech** [reply]
It should just throw a MethodNotFound if you try and reflect the method call and it doesn't exist. It's potentially expensive though.

**Reply by dantleech** [reply]
Also it doesn't matter if the class is in the current namespace or not - the existing code transform will generate the method in the whatever class it belongs to, might need some work to make that work with the LS though

**Reply by BladeMF** [reply]
Is there any other method of doing this that you can think of?

**Reply by BladeMF** [reply]
I meant "workspace", rather than "namespace". By "some work" do you mean that current GenerateMethod implementation needs to migrate from `SourceCode` to `TextEdits` return value?

Otherwise that check for undefined methods is probably unavoidable. Interestingly, phpstan does the same thing.

**Reply by dantleech** [reply]
Yes -- we can add an `lint` method to the `SoureCodeReflector` which returns all the issues (most issues can be picked up when the code is statically analyzed), and there is already some unused functionality around this:

https://github.com/phpactor/worse-reflection/blob/master/lib/Core/Inference/SymbolContextResolver.php#L265

But that would need to be refactored to collect all errors `Diagnostic($code, $offset, $message)`.

pff...

**Reply by BladeMF** [reply]
I thought I picked something easy...

**Reply by dantleech** [reply]
I might have a go at creating the Lint method - would be quite useful for providing basic diagnostics etc

**Reply by BladeMF** [reply]
I will try to implement the actual code action with the only restriction of appearing on method calls - just to get things started. I'd rather have all the refactorings **somewhat** implemented, rather than not have them. I believe that people who use phpactor would agree.

**Reply by dantleech** [reply]
Right, have implemented but it's not very useful, so not merging. WR is such a huge mess of technical debt... in order to support generics much of it needs to be rewritten from scratch.

Maybe the best option for now would be indeed to brute force it as you originally suggested. WR uses a cache, so it might not have so much impact.

**Reply by BladeMF** [reply]
Let me show you what I mean in a PR and then we'll discuss.
